src/burnchains/affirmation.rs

```diff
@@ -0,0 +1,877 @@
+/// commit is found, return the descendancy matrix for it as well.
+/// Returns Some(the winning block commit, descendancy matrix, total confirmations, total burnt) if
+/// there's an anchor block commit.
+/// Returns None otherwise
+pub fn find_heaviest_block_commit<'a, B: BurnchainHeaderReader>(
+    burnchain_tx: &BurnchainDBTransaction<'a>,
+    indexer: &B,
+    prepare_phase_ops: &Vec<Vec<LeaderBlockCommitOp>>,
+    anchor_threshold: u32,
+) -> Result<Option<(LeaderBlockCommitOp, Vec<Vec<bool>>, u64, u64)>, DBError> {
+    // sanity check -- must be in order by block height and vtxindex
+    for prepare_block_ops in prepare_phase_ops.iter() {
+        let mut expected_block_height = None;
+        let mut last_vtxindex = None;
+        for opdata in prepare_block_ops.iter() {
+            if let Some(expected_block_height) = expected_block_height.as_ref() {
+                assert_eq!(expected_block_height, &opdata.block_height);
+            } else {
+                expected_block_height = Some(opdata.block_height);
+            }
+
+            if let Some(last_vtxindex) = last_vtxindex.as_mut() {
+                assert!(*last_vtxindex < opdata.vtxindex);
+                *last_vtxindex = opdata.vtxindex;
+            } else {
+                last_vtxindex = Some(opdata.vtxindex);
+            }
+            test_debug!(
+                "Prepare-phase block-commit {},{}: {}",
+                opdata.block_height,
+                opdata.vtxindex,
+                &opdata.txid
+            );
+        }
+    }
+
+    // map (block_height, vtxindex) to (parent_block_height, parent_vtxindex)
+    let mut parents = BTreeMap::new();
+
+    // map (block_height, vtxindex) to (non-prepare-ancestor-height, non-prepare-ancestor-vtxindex)
+    let mut ancestors = BTreeMap::new();
+
+    // map (non-prepare-ancestor-height, non-prepare-ancestor-vtxindex) to (set-of-block-heights, total_burnt)
+    // that contain descendants
+    let mut ancestor_confirmations: BTreeMap<(u64, u32), (HashSet<u64>, u64)> = BTreeMap::new();
+
+    // calculate each block-commit's parents
+    for prepare_block_ops in prepare_phase_ops.iter() {
+        for opdata in prepare_block_ops.iter() {
+            parents.insert(
+                (opdata.block_height, opdata.vtxindex),
+                (
+                    opdata.parent_block_ptr as u64,
+                    opdata.parent_vtxindex as u32,
+                ),
+            );
+        }
+    }
+
+    // calculate the ancestor map -- find the highest non-prepare-phase ancestor for each prepare-phase block-commit.
+    for prepare_block_ops in prepare_phase_ops.iter().rev() {
+        for opdata in prepare_block_ops.iter() {
+            let mut cursor = (opdata.block_height, opdata.vtxindex);
+            while let Some((parent_block, parent_vtxindex)) = parents.get(&cursor) {
+                cursor = (*parent_block, *parent_vtxindex);
```

Comment:
This can be made slightly more efficient (and less quadratic) by checking if the cursor is already in the ancestors map, right?

Reply:
Yup, done